### PR TITLE
Feature: Email Validation with Mailosaur

### PR DESCRIPTION
**Pull request by m2rads**
This PR addressed issue: #104 

### Added  
- Added support for Mailosaur email validation
- Added email rendering feature in the browser
- Added sleep_milliseconds tool to add delays in the test execution
- Added more robust error handling for Mailosaur email validation

Demo: 

https://github.com/user-attachments/assets/a45ac3ad-27f4-4dcc-825a-a5192d62cee4

### CODE REVIEW

File: packages/shortest/CHANGELOG.md [hunk 1]
```diff
@@ -5,6 +5,14 @@ All notable changes to this project will be documented in this file.
...
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [0.1.2] - 2024-12-26
+
+### Added  
+- Added support for Mailosaur email validation
+- Added email rendering feature in the browser
+- Added sleep_milliseconds tool to add delays in the test execution
+- Added more robust error handling for Mailosaur email validation
+
```
Comment (m2rads) [inline packages/shortest/CHANGELOG.md:15]: Updated the change logs to notify the end user about the changes in this version. Used the wording "Added" to imply that this version includes new features. Maybe we can come up with a better wording convention for our change logs to stay consistent.

File: packages/shortest/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@antiwork/shortest",
-  "version": "0.1.1",
+  "version": "0.1.2",
```
Comment (m2rads) [inline packages/shortest/package.json:3]: Very important - Every change for the package  requires version bump in order to publish to NPM successfully. Otherwise our publish workflow will fail. Example of a failure: 

image

File: packages/shortest/package.json [hunk 3]
```diff
@@ -29,36 +29,38 @@
...
-    "build:cli": "esbuild src/cli/bin.ts src/cli/setup.ts --bundle --platform=node --format=esm --outdir=dist/cli --metafile=dist/meta-cli.json --external:fsevents --external:chokidar --external:glob --external:esbuild --external:events --external:path --external:fs --external:util --external:stream --external:os --external:assert --external:url --external:playwright --external:@anthropic-ai/sdk --external:expect --external:dotenv --external:otplib --external:picocolors --external:punycode",
+    "build:cli": "esbuild src/cli/bin.ts src/cli/setup.ts --bundle --platform=node --format=esm --outdir=dist/cli --metafile=dist/meta-cli.json --external:fsevents --external:chokidar --external:glob --external:esbuild --external:events --external:path --external:fs --external:util --external:stream --external:os --external:assert --external:url --external:playwright --external:@anthropic-ai/sdk --external:expect --external:dotenv --external:otplib --external:picocolors --external:punycode --external:https --external:http --external:net --external:tls --external:crypto --external:mailosaur",
```
Comment (m2rads) [inline packages/shortest/package.json:32]: Mailosaur requires these Node Built-in modules to be marked as external to work. Doing this will reduce the package size. 

https
http
net
tls
crypto 
mailosaur

File: packages/shortest/src/ai/prompts/index.ts [hunk 4]
```diff
@@ -11,9 +11,10 @@ Expect: 1. Test case to be generated within at least 20 seconds [HAS_CALLBACK]
 IMPORTANT GLOBAL RULES:
 
 1. **Waiting for Conditions**:
-   - Some steps will require waiting before proceeding to the next action.
-   - This waiting can be based on a time delay (e.g., seconds or minutes) or waiting for an element to become visible or clickable.
-   - If the specified condition is not met after the allotted time, the test should be considered failed.
+   - When you are specifically asked to wait for a condition, use the "sleep" tool. Otherwise, we internally handle proper waiting for elements to load.
+   - For time-based waits:
+     - Use { action: "sleep", duration: X } where X is milliseconds
+   - Always wait for the tool to finish before proceeding to the next action. You will recieve a message to continue with your next action once the wait is over. Then validate the condition is met.
```
Comment (m2rads) [inline packages/shortest/src/ai/prompts/index.ts:17]: Improving the prompt for the "sleep" tool to instruct AI to use this tool whenever wait is needed to fulfill a condition.

File: packages/shortest/src/ai/prompts/index.ts [hunk 5]
```diff
@@ -42,6 +43,14 @@ IMPORTANT GLOBAL RULES:
    - All expectations listed in the test instructions must be fulfilled.
    - If any expectation is not met, the test case must be marked as failed.
 
+8. **Testing Email**:
+   - If you need to test a condition that involves seeing the contents of an email, use the "check_email" tool.
+   - For email validation, you MUST always use 'Click' and 'Mouse' action instead of using keyboard shortcuts.
+   - This tool will grab the latest email from the email address given to you and will render it in a new tab for you to see.
+   - Once you are done with validating the email, navigate back to the original tab.
+   - You MUST pass the email address that is given to you to the tool as a parameter otherwise it will fail.
+   - If no email address is given to you for this test, you should fail the test.
```
Comment (m2rads) [inline packages/shortest/src/ai/prompts/index.ts:52]: Added a new prompt rule for Email Validation. Tried to stay consistent with the convention we have in place. 

We explicitly ask AI to use Click and Mouse events because sometimes AI might use tabs to focus on the next text filed that might result in focusing on something else. This would make the AI more reliable.

File: packages/shortest/src/ai/tools/index.ts [hunk 6]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
...
+      properties: {
+        action: {
+          type: "string",
+          enum: ["check_email"],
+          description: "Render the received email in a new tab"
+        }
+      },
+      required: ["action", "email"]
+    }
+  },
```
Comment (m2rads) [inline packages/shortest/src/ai/tools/index.ts:47]: In this file we register our tools. We have created a new tool called "check_email". It let's AI know what we expect from it when calling this tool. The required field at the bottom asks AI to provide the email address as an ActionInput.

File: packages/shortest/src/ai/tools/index.ts [hunk 7]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
...
+        duration: {
+          type: "number",
+          description: "Duration to sleep in milliseconds (e.g. 5000 for 5 seconds)",
+          minimum: 0,
+          maximum: 60000
+        }
+      },
+      required: ["action", "duration"]
+    }
+  },
```
Comment (m2rads) [inline packages/shortest/src/ai/tools/index.ts:68]: Another tool that we needed to ship with "check_email" is "sleep" tool. This tool helps AI to wait for a condition to be met before performing any other actions. We needed this tool because seems like AI does not have any perception of time so we implemented this to be more reliable with waiting. 


Why we need both of these tools in the same PR? 

Because sometimes we might need to wait for the Email to be received from the client.

File: packages/shortest/src/browser/integrations/index.ts [hunk 8]
```diff
@@ -1,3 +1,2 @@
-import { GitHubTool } from "./github";
-
-export { GitHubTool };
+export { GitHubTool } from './github';
+export { MailosaurTool } from './mailosaur';
```
Comment (m2rads) [inline packages/shortest/src/browser/integrations/index.ts:2]: index file to export all the integrations that will be used in `browser-tools.ts`

File: packages/shortest/src/browser/integrations/mailosaur.ts
Comment (m2rads) [inline packages/shortest/src/browser/integrations/mailosaur.ts:1]: This file is the integration for Mailosaur. It used Milosaur SDK to initialize and get the last email received from the specified email address that is passed to this class by AI.

File: packages/shortest/src/types/browser.ts [hunk 9]
```diff
@@ -27,7 +27,9 @@ export type BrowserAction =
   | "type"
   | "key"
   | "run_callback"
-  | "navigate";
+  | "navigate"
+  | "sleep"
+  | "check_email";
```
Comment (m2rads) [inline packages/shortest/src/types/browser.ts:32]: Adding these types to the list of our Browser actions. This will help us to validate the AI will not send an action that we have not implemented yet.

File: packages/shortest/src/types/browser.ts [hunk 10]
```diff
@@ -43,6 +45,8 @@ export interface ActionInput {
   username?: string;
   password?: string;
   url?: string;
+  duration?: number;
+  email?: string;
```
Comment (m2rads) [inline packages/shortest/src/types/browser.ts:49]: these are action input types. Helps us to validate that AI is sending the required inputs.

File: packages/shortest/src/types/config.ts [hunk 11]
```diff
@@ -6,4 +6,8 @@ export interface ShortestConfig {
   baseUrl: string;
   testDir: string | string[];
   anthropicKey: string;
+  mailosaur?: {
+    apiKey: string;
+    serverId: string;
+  };
```
Comment (m2rads) [inline packages/shortest/src/types/config.ts:12]: In order to use Email validation feature, end users need to add these configurations.

File: packages/shortest/tests/test-email.ts
Comment (m2rads) [inline packages/shortest/tests/test-email.ts:1]: A simple test script to mock the email validation functionality when testing.

File: shortest.config.ts
Comment (m2rads) [inline shortest.config.ts:1]: Updated the config file to perform email validation

**Review by m2rads — COMMENTED** [review]
@slavingia, @rmarescu 

I added self-review as requested. If there is any part that is vague or requires more explanation please let me know.

File: app/__tests__/dashboard.test.ts [hunk 12]
```diff
@@ -1,21 +1,4 @@
...
-    signInParams: {
-      strategy: "email_code",
-      identifier: "shortest+clerk_test@example.com",
-    },
-  });
-
-  await page.goto(frontendUrl + "/dashboard");
-});
 
+shortest("Login to the App using magic link. Use this email: 'stay-ants@taoyeoeo.mailosaur.net'");
```
Comment (m2rads) [inline app/__tests__/dashboard.test.ts:3]: Removing injecting session into browser and instead using magic link. Because this would perform email validation as well as login validation. 

Here's a demo: 


https://github.com/user-attachments/assets/160f3183-49fe-406c-a781-6124cf77676c

File: packages/shortest/CHANGELOG.md (hunk 1 above)
Comment (rmarescu) [reply packages/shortest/CHANGELOG.md:15]: Seems redundant to use `Added` for each bullet point when all the changes are of type `Added`. Also, the bullet points should use present tense.

What about this?

```suggestion
## [0.1.2] - 2024-12-26

### Added
- Mailosaur integration with error handling for email validation 
- Browser-based email preview functionality
- Test execution delay utility (sleep_milliseconds)
```

File: packages/shortest/src/ai/prompts/index.ts (hunk 4 above)
Comment (rmarescu) [inline packages/shortest/src/ai/prompts/index.ts:17]:
```suggestion
   - Always wait for the tool to finish before proceeding to the next action. You will receive a message to continue with your next action once the wait is over. Then validate the condition is met.
```

File: packages/shortest/src/ai/prompts/index.ts [hunk 13]
```diff
@@ -42,6 +43,14 @@ IMPORTANT GLOBAL RULES:
    - All expectations listed in the test instructions must be fulfilled.
    - If any expectation is not met, the test case must be marked as failed.
 
+8. **Testing Email**:
+   - If you need to test a condition that involves seeing the contents of an email, use the "check_email" tool.
+   - For email validation, you MUST always use 'Click' and 'Mouse' action instead of using keyboard shortcuts.
+   - This tool will grab the latest email from the email address given to you and will render it in a new tab for you to see.
+   - Once you are done with validating the email, navigate back to the original tab.
```
Comment (rmarescu) [inline packages/shortest/src/ai/prompts/index.ts:50]: It would close the email preview tab automatically?

File: packages/shortest/src/ai/tools/index.ts [hunk 14]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
       required: ["action", "username", "password"],
     },
   },
+  {
+    name: "check_email",
```
Comment (rmarescu) [inline packages/shortest/src/ai/tools/index.ts:34]: What about 

```suggestion
    name: "view_email",
```

File: packages/shortest/src/ai/tools/index.ts [hunk 15]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
...
+    description: "Fetch, decode and render received email in browser",
```
Comment (rmarescu) [inline packages/shortest/src/ai/tools/index.ts:35]: Fetch / decode are implementation details. 

```suggestion
    description: "View received email in new browser tab",
```

File: packages/shortest/src/ai/tools/index.ts [hunk 16]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
...
+    input_schema: {
+      type: "object",
+      properties: {
+        action: {
+          type: "string",
+          enum: ["check_email"],
```
Comment (rmarescu) [inline packages/shortest/src/ai/tools/index.ts:41]:
```suggestion
          enum: ["view_email"],
```

File: packages/shortest/src/ai/tools/index.ts [hunk 17]
```diff
@@ -30,6 +30,42 @@ export const AITools = [
...
+          description: "Render the received email in a new tab"
```
Comment (rmarescu) [inline packages/shortest/src/ai/tools/index.ts:42]: Not familiar yet on why the action needs to be repeated, but can it have the same description, for consistency?

```suggestion
          description: "View received email in new browser tab"
```

File: packages/shortest/src/browser/core/browser-tool.ts [hunk 18]
```diff
@@ -39,64 +35,65 @@ export class BrowserTool extends BaseBrowserTool {
...
+    page: Page, 
     browserManager: BrowserManager,
-    config: BrowserToolConfig,
+    config: BrowserToolConfig
   ) {
     super(config);
     this.page = page;
     this.browserManager = browserManager;
-    this.screenshotDir = join(process.cwd(), ".shortest", "screenshots");
+    this.screenshotDir = join(process.cwd(), '.shortest', 'screenshots');
```
Comment (rmarescu) [inline packages/shortest/src/browser/core/browser-tool.ts:49]: Any linting changes should be done in a separate PR; otherwise, it increased the diff for this one, and it makes it harder to review.

**Review by rmarescu — COMMENTED** [review]
I've added some initial comments, haven't had a chance to go over all the changes yet. Will continue tomorrow.

File: packages/shortest/src/ai/tools/index.ts (hunk 14 above)
Comment (slavingia) [reply packages/shortest/src/ai/tools/index.ts:34]: I prefer check (this is why the same PR would be good)

File: app/__tests__/dashboard.test.ts (hunk 12 above)
Comment (slavingia) [inline app/__tests__/dashboard.test.ts:3]:
```suggestion
shortest("Login to the app using a magic link with this email: stay-ants@taoyeoeo.mailosaur.net");
```

File: packages/shortest/CHANGELOG.md (hunk 1 above)
Comment (slavingia) [reply packages/shortest/CHANGELOG.md:15]: Agreed ^

File: packages/shortest/src/ai/tools/index.ts (hunk 17 above)
Comment (slavingia) [inline packages/shortest/src/ai/tools/index.ts:42]:
```suggestion
          description: "Check that the email was received with specified content in a new tab"
```

File: packages/shortest/src/ai/prompts/index.ts (hunk 13 above)
Comment (m2rads) [reply packages/shortest/src/ai/prompts/index.ts:50]: No it does not but it will change the browser tab to the origin.

File: packages/shortest/src/ai/tools/index.ts (hunk 17 above)
Comment (m2rads) [reply packages/shortest/src/ai/tools/index.ts]: We don't need to have this again actually but we can use it in case we'd like to give AI more context. Should I remove this you think?

File: packages/shortest/src/browser/core/browser-tool.ts (hunk 18 above)
Comment (m2rads) [reply packages/shortest/src/browser/core/browser-tool.ts]: I fixed the linting errors in a different branch and merged main here. Autofix is passing!

File: app/__tests__/dashboard.test.ts (hunk 12 above)
Comment (m2rads) [reply app/__tests__/dashboard.test.ts]: Changed it to the one you sent me. `mission-health@rdt7stzf.mailosaur.net`

File: packages/shortest/src/ai/tools/index.ts (hunk 17 above)
Comment (m2rads) [reply packages/shortest/src/ai/tools/index.ts]: Updated!

File: packages/shortest/src/ai/prompts/index.ts (hunk 4 above)
Comment (m2rads) [reply packages/shortest/src/ai/prompts/index.ts:14]: Addressed!

File: app/__tests__/dashboard.test.ts (hunk 12 above)
Comment (slavingia) [reply app/__tests__/dashboard.test.ts]: Sounds good but please make the other changes, I think it makes it more human-readable / a better example once AI is used to generate new shortest tests for other repos

**Review by slavingia — APPROVED** [review]

File: packages/shortest/src/ai/tools/index.ts (hunk 17 above)
Comment (m2rads) [reply packages/shortest/src/ai/tools/index.ts]: I added Sahil's recommendation for Action description and yours for tool description. I think this would give AI more context and help it to be more consistent.

File: packages/shortest/package.json [hunk 19]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@antiwork/shortest",
-  "version": "0.1.1",
+  "version": "0.2.1",
```
Comment (m2rads) [inline packages/shortest/package.json:3]: Changed the version to 0.2.1 to indicate that this version introduces a new feature. This would be more consistent with semantic versioning best practices. 

image
